src/embed/connection.rs

```diff
@@ -186,30 +156,62 @@ impl<'a, 'b> Context<'a, 'b> {
     }
 }
 
-async fn execute<'a, 'b, T, S, F, E>(ctx: Context<'a, 'b>, func: F, req: Request)
-where
-    T: serde::de::DeserializeOwned,
-    S: serde::Serialize,
-    E: std::fmt::Debug,
-    F: AsyncFn2<Context<'a, 'b>, T, Output = Result<S, E>>,
-{
-    let request: T = match deserialize(&req.data.as_string().unwrap()) {
-        Ok(v) => v,
-        Err(e) => return req.response.send(Err(e)).await,
-    };
+#[derive(Debug)]
+enum ExecuteError {
+    TransactionNotFound(u32),
+    TransactionRequired,
+    TransactionIsReadOnly(u32),
+}
 
-    let result = func
-        .call(ctx, request)
-        .await
-        .map(|v| serde_json::to_string(&v));
-    // When https://doc.rust-lang.org/std/result/enum.Result.html lands we can remove this.
-    let result = match result {
-        Ok(Ok(v)) => Ok(v),
-        Ok(Err(v)) => Err(to_debug(v)),
-        Err(v) => Err(to_debug(v)),
+async fn execute<'a, 'b>(ctx: Context<'a, 'b>, req: &Request) -> Result<String, String> {
+    use ExecuteError::*;
+
+    // transaction-less
+    match req.rpc.as_str() {
+        "getRoot" => return to_json(do_get_root(ctx, from_json(&req.data)?).await),
+        "openTransaction" => return to_json(do_open_transaction(ctx, from_json(&req.data)?).await),
+        "commitTransaction" => return to_json(do_commit(ctx, from_json(&req.data)?).await),
+        "closeTransaction" => return to_json(do_close(ctx, from_json(&req.data)?).await),
+        "beginSync" => return to_json(do_begin_sync(ctx, from_json(&req.data)?).await),
+        "maybeEndSync" => return to_json(do_maybe_end_sync(ctx, from_json(&req.data)?).await),
+        _ => (),
     };
 
-    req.response.send(result).await
+    // require read txn
+    let txn_req: TransactionRequest = from_json(&req.data)?;
+    let txn_id = txn_req
+        .transaction_id
+        .ok_or(TransactionRequired)
+        .map_err(to_debug)?;
+    let txn_id_string = txn_id.to_string();
+    req.lc.add_context("txid", &txn_id_string);
+    let txns = ctx.txns.read().await;
```

Comment:
This is how it was before, but this bit concerns me. It looks to me like an addition to the transaction map (e.g., `openTransaction`, even a read-only transaction) requires waiting for all other transactions to close?

Reply:
Hmmm. I'm not sure what the intent was. I think the intent might have been that txns mutex just protects the map of transaction to id, in which case anyone who acquires it should do their thing (read or insert or remove or whatever) and immediately drop the lock. There is at least one other place besides here where that is not happening.

That intent would not cover ensuring that only one dispatch call was processing a given transaction at a time, which would require I think a lock per transaction. Not sure if that is important. 

This seems like something we need to discuss more.

Reply:
https://github.com/rocicorp/repc/issues/240